### ISSUE TITLE
Download speed is very low if torrent client is working on upload and download

### ISSUE THREAD

**Issue by shamarin** [AdguardTeam/AdGuardVPNForWindows#726]
### Please answer the following questions for yourself before submitting an issue.

- [X] I checked the [knowledge base](https://adguard-vpn.com/kb/) and found no answer
- [X] I checked to make sure that this issue has not already been filed

### AdGuard VPN version

2.4 nightly 11

### Environment

```markdown
- OS: Win 11 x64 22635.4145
- Other network-level software:
```

### Which DNS server do you use?

Default

### Use Wintun

- [ ] yes, I do

### Issue Details

Steps to reproduce:
1. Start torrent client with active seeding
2. Open Adguard VPN and connect to any fast VPN server
3. Go to speedtest.net and check or open any online video and download speed will be maximum 2-5 Mbits, upload about 20-30 Mbits

### Expected Behavior

Download speed should be about 20-30 Mbits as for 100 Mbit ISP channel. Also torrent client was confirued so that it utilizes only 70 Mbits from 100 Mbits channel width.

### Actual Behavior

Download speed is very low.

### Screenshots

<details><summary>Screenshot 1</summary>

</details>

### Additional Information

torrent client was confirued so that it utilizes only 70 Mbits from 100 Mbits channel width. I use Qbittorrent client for torrents.

**Comment by shamarin** [comment]
![изображение](https://github.com/user-attachments/assets/c528a1e9-485c-4092-9174-8ef9cce057c8)

**Comment by shamarin** [comment]
![изображение](https://github.com/user-attachments/assets/a85c9800-1add-41d0-a330-34edd22cfb7b)

**Comment by shamarin** [comment]
Link to the logs https://cloud.mail.ru/public/SLz9/qRfURjH8T
It was reproduced at 31.08.2024 13:08-13:11 Moscow time

**Comment by shamarin** [comment]
With real app (FDM download manager)
![изображение](https://github.com/user-attachments/assets/827851ca-a261-44f3-ba87-dccd77041dfc)

**Comment by shamarin** [comment]
The same but without Adguard VPN connection 
![изображение](https://github.com/user-attachments/assets/153d52db-2ed0-4079-baee-7926b4c5fb54)

**Comment by shamarin** [comment]
With Kaspersky VPN
![изображение](https://github.com/user-attachments/assets/841a6db5-cc7e-4803-bb9d-7b26fd42caa2)

**Comment by shamarin** [comment]
The same conditions but Adguard VPN is very slow in that condition.

**Comment by shamarin** [comment]
Torrent client also is in the exlusions in both Adguard and Kaspersky.

**Comment by KolbasovAnton** [comment]
@shamarin Greetings! On the newest nightly version (2.4.0 nightly 12) situation didn't get any better?

**Comment by shamarin** [comment]
> @shamarin Greetings! On the newest nightly version (2.4.0 nightly 12) situation didn't get any better?

Better, but some more tuning needed.

**Comment by shamarin** [comment]
> @shamarin Greetings! On the newest nightly version (2.4.0 nightly 12) situation didn't get any better?

Tested for some more time, not better with new nightly version. Speed is very slow if to compare with Kaspersky for example. 
Maybee situation will be better when Wintun high CPU load issue will be fixed and I will test it with Wintun enabled.

**Comment by shamarin** [comment]
One thing - at this moment always ping from VPN servers is high if to press refresh button in Adguard, if again to press it after some seconds ping will be lower. For example for Sweeden as fast VPN normal ping is about 30-33 ms, if to press update then it will be 95 ms and back again after some seconds. Maybee in this is the problem??? it happens for many other VPN locations.

**Comment by shamarin** [comment]
> @shamarin Greetings! On the newest nightly version (2.4.0 nightly 12) situation didn't get any better?

With the latest nightly version 2.4.0 12 Youtube speed is low with Wintun or without it and without working torrent client, QUIC adds some speed. So, 4k videos stucks on buffering, max speed is about 10-20 Mbits with any fast location (I use mostly Denmark, Deutchland, Sweeden or Nothland).
Youtube is blocked in Russia (traffic from Youtube is degraded by DPI).

**Comment by KolbasovAnton** [comment]
@shamarin Hello! Any progress on newest nightly (2.4.1 n 3)? Just now I was testing your case with this version and get ~25 Mb downloading speed on 11 seeds. Qbittorrent was also in general exclusions, as You wrote [here](https://github.com/AdguardTeam/AdGuardVPNForWindows/issues/726#issuecomment-2322950027).

**Comment by shamarin** [comment]
> @shamarin Hello! Any progress on newest nightly (2.4.1 n 3)? Just now I was testing your case with this version and get ~25 Mb downloading speed on 11 seeds. Qbittorrent was also in general exclusions, as You wrote [here](https://github.com/AdguardTeam/AdGuardVPNForWindows/issues/726#issuecomment-2322950027).

Hello! No, just the same (I couldn't get 25 Mbit on VPN traffic, my seeds is higher than 11). Will test today with latest test build from #693
Label "Waiting for data" means that you are waiting from me a new logs?
My older logs  https://cloud.mail.ru/public/SLz9/qRfURjH8T from August is not enough?

**Comment by KolbasovAnton** [comment]
@shamarin Yes, it would be great to reproduce it once again on latest builds. The more recent the logs, the better.

**Comment by shamarin** [comment]
> @shamarin Yes, it would be great to reproduce it once again on latest builds. The more recent the logs, the better.

Протестировал с последней тестовой сборкой от команды разработчиков согласно случая #693 . В самом начале скорость соединения 20 Мбит/ 70 Мбит (54 торрент раздачи активно одновременно добавлено). Подключение ПК проводное (по Wi-Fi значительно лучше всегда VPN работает почему то - максимальные скорости без работающего торрент клиента). Максимальная ширина канала интернет 100 Мбит, маршрутизатор Keenetic Peak. Спустя пару минут после того как запущен торрент клиент и VPN подключено, то скорость падает до 3-4 Мбит по низсходящему каналу, по восходящему 40-50 Мбит. Такое ощущение что канал забивает чем-то. 
<img width="221" alt="Снимок экрана 2024-10-18 175113" src="https://github.com/user-attachments/assets/a79c4097-74a5-4b24-be25-0e8a415d4e14">
![Screenshot 2024-10-18 at 18-05-38 Speedtest by Ookla - The Global Broadband Speed Test](https://github.com/user-attachments/assets/bccc59ac-da7b-4729-8e0a-ac553f8f2fa4)
Логи снял  https://cloud.mail.ru/public/Wm4h/XjNrwT27g (время воспроизведения проблемы с 18:04-18:09)

**Comment by shamarin** [comment]
Так-то вообще очень странно, выключил VPN и торрент клиент закрыл полностью. Скорость замерил с теми же параметрами
![изображение](https://github.com/user-attachments/assets/f6bc9559-6fa4-4ac4-b554-b7fc237a6c1c)

**Comment by shamarin** [comment]
Германия (Франкфкрт) с включенным VPN
![изображение](https://github.com/user-attachments/assets/3d8dda88-c1ed-4e08-b2cc-ac251bb7ebf0)

**Comment by shamarin** [comment]
С выключенным VPN на том же IP адресе
![изображение](https://github.com/user-attachments/assets/c8558f8f-3574-4874-95f6-6ff960910bf6)
Во всех случаях VPN Wintun включен. Заметил, что в последних версиях ночных с Wintun скорость нисходящая стала реально низкой - 30-40 Мбит. По Wi-Fi 100 Мбит выдает и туда и туда. Логи собирал с включенным Wintun.

**Comment by shamarin** [comment]
IP Германии (Франкфурт), но подключился к локации Дании
![изображение](https://github.com/user-attachments/assets/4dcd9277-6b4d-440f-82ce-52f46d556219)
В общем, что-то не так у вас с определением ширины канала нисходящего. И в целом всё это огорчает, сервис вроде хороший, но слишком сырой - Wintun с торрентом не возможно использовать и до сих пор не исправлено с лета ошибка, потом вот такая вот проблема когда торрент работает, то не возможно видео смотреть. Плюс еще и без торрента скорость нисходящая низкая. А если QUIC выключить, то скорость на Youtube совсем низкая становится - 2-3 Мбит и там даже с 1080P видео смотреть не возможно. Вот и думаю стоит ли продлевать лицензию, так как продукт слишком сырой и ошибки очень медленно исправляются.

**Comment by shamarin** [comment]
@KolbasovAnton когда уже наконец-то разработчики возьмутся за данную проблему и исправят её???

**Comment by shamarin** [comment]
Иногда без работающего торрента такая низкая скорость нисходящая.
![изображение](https://github.com/user-attachments/assets/c1d94d77-2db5-47d1-945c-f6a3c95335ec)

**Comment by shamarin** [comment]
Воспроизвел и логи собрал (время с 22:11 - 22:13)

**Comment by shamarin** [comment]
https://cloud.mail.ru/public/FeD3/r8CkvcNet

**Comment by shamarin** [comment]
И приложение показывает не верные задержки для локаций. Можно сравнить ту же Данию в speedtest и что показывает приложение. Приложение показывает:
![изображение](https://github.com/user-attachments/assets/0532f30a-18bc-4c71-852d-79feb88de113)
А speedtest показывает 100 мс в среднем.

**Comment by shamarin** [comment]
и для сравнения Kaspersky Secure Connection, та же локация Дания и протокол Wireguard Wintun 
![изображение](https://github.com/user-attachments/assets/2f5155b7-b011-4e4b-b61c-e926e3916428)
Разница существенна.

**Comment by shamarin** [comment]
Browsec (браузерное расширение) для локации Франция
![изображение](https://github.com/user-attachments/assets/215903b9-2adb-4487-8159-fc82d03d5ad1)

**Comment by shamarin** [comment]
На ноутбуку по Wi-Fi, оператор ТТК
![изображение](https://github.com/user-attachments/assets/cefc0508-573a-4e85-976e-073c1ac2336d)

**Comment by shamarin** [comment]
![изображение](https://github.com/user-attachments/assets/bec30e3a-330e-4558-bfd6-e3509f3751a3)

**Comment by shamarin** [comment]
> Иногда без работающего торрента такая низкая скорость нисходящая. ![изображение](https://private-user-images.githubusercontent.com/24771090/377985141-c1d94d77-2db5-47d1-945c-f6a3c95335ec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk0MzA5ODIsIm5iZiI6MTcyOTQzMDY4MiwicGF0aCI6Ii8yNDc3MTA5MC8zNzc5ODUxNDEtYzFkOTRkNzctMmRiNS00N2QxLTk0NWMtZjZhM2M5NTMzNWVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIwVDEzMjQ0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc2MjBkNjM4MjAyYzI0MTRjMWY5NmMxOTRjZjczNzAxZmEyNjE3NzlmZmZlNzY3NjIxMjg1MzAxNmIwMmFkYzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-2dI6nqWyTnkXRxqqn939G6lq8-vJiCg4uLM7dlwfXk)

Почему-то только у оператора Ростелеком такая проблема у меня возникает. Проверял и по Wi-Fi соединению, там тоже выше 33 Мбит не выдает скорость нисходящего потока. У ТТК и у Дом.Ру все как положено - скорость высокая. У Дом.Ру по тарифу 300 Мбит, с VPN получил нисходящую 200-233 Мбит, что очень хорошо. Может оператор Ростелеком что-то блокирует у вас?

**Comment by shamarin** [comment]
Если Wintun отключить и торрент клиент убрать из исключений, то тогда скорость высокая и в торрент клиенте и на остальных ресурсах. В общем у вас какая то проблема с прохождением трафика от торрент клиента. Такое ощущение, что Adguard вообще игнорирует правило исключения для торрент клиента и весь трафик в торрент клиенте идет все равно напрямую.

**Comment by shamarin** [comment]
И с Wintun есть проблема очень серьезная (описано в #693 ) и она до сих пор не исправлена, хотя я указывал еще в середине июня о ней, в июле взяли в работу и так до сих пор не исправили.

**Comment by shamarin** [comment]
![изображение](https://github.com/user-attachments/assets/c490954d-6ad8-467e-8fba-765b68c71254)

**Comment by shamarin** [comment]
![изображение](https://github.com/user-attachments/assets/b1dd62f4-455d-42d8-899d-858e9f02e86b)

**Comment by shamarin** [comment]
![изображение](https://github.com/user-attachments/assets/b9f5873a-b6d6-4a44-99f2-83942353fc27)

**Comment by shamarin** [comment]
![изображение](https://github.com/user-attachments/assets/914893e3-0359-4ff3-81f7-21ca5695082c)

**Comment by shamarin** [comment]
![изображение](https://github.com/user-attachments/assets/6d5dbea6-7afe-4bbe-8d52-36b92e53511c)

**Comment by shamarin** [comment]
В статистике данных о торрент клиенте как потребителе трафика информация отсутствует. Значит трафик идет напрямую.

**Comment by shamarin** [comment]
Высокая скорость на других ресурсах сохраняется максимум пару минут и потом опять в районе 4-5 Мбит максимум становится. Торрент клиент при этом на максималках скорость (70 Мбит).

**Comment by shamarin** [comment]
Спустя очень большое время в статистике qbittorrent появился, но там всего в районе 160 МБ данных в оба направления, хотя за это время пару гигабайт должно быть. В общем странно работает пропуск трафика от торрент клиентов.

**Comment by shamarin** [comment]
@KolbasovAnton при 11 сидах (11 раздач) скорость действительно приемлемая, но даже уже при 12 раздачах резко падает скорость. Торрент клиент в исключениях.

**Comment by shamarin** [comment]
Сегодня вышло обновление 2.4.1 nightly 4. С Wintun исправлено, скорость в начале подольше стала держаться высокая. Потом спустя минут 5-10 низка становится. Использую Wintun и QUIC, на раздаче 14 активных торрентов.
Сначала подключаюсь к VPN потом запускаю торрент клиент. В начале еще особенность - диспетчер задач Windows показывает что Adguard VPN использует активно сеть (60-70 Мбит), но при этом торрент клиент в исключениях и других источников обращения к сети в данный момент нет. Вот когда спустя время в диспетчере уже перестает показывает это, то тогда скорость VPN соединения низкая становится.

**Comment by shamarin** [comment]
Может лучше реализовать поддержку Wireguard на альтернативных и более современных решениях - https://github.com/cloudflare/boringtun или https://git.zx2c4.com/wireguard-nt/about/ в режиме kernel. Судя по всему там производительность выше.

**Comment by shamarin** [comment]
@KolbasovAnton мне не понятно только поведение исключения трафика торрент клиента Adguard VPN.
Путь воспроизведения проблемы:
1. Подключиться к любой локации Adguard VPN
2. Торрент клиент не запущен, другие потребители трафика тоже не активны. Убедиться в этом через диспетчер задач Windows посмотрев напротив запущенного приложения Adguard VPN. Оно трафик не потребляет. 
3. Запустить торрент клиент с активными раздачами.
4. Как только торрент клиент начинает использовать сеть, то хоть оно и в исключениях Adguard, то Adgurad VPN потреблять трафик начинает порядка 40 Мбит, иногда больше.
Может в этом и причина что c включенным Wintun потребление системных ресурсов (процессора) приложением становится огромным и плюс скорость соединения (хоть и должна быть полоса порядка 40 Мбит доступна, так как для торрент клиента сделано ограничение в 60 Мбит максимум и ограничение по служебному трафику и по трафику UDP тоже и для локальных пиров) становится рекордно низкой, что с Wintun, что без него (WFP драйвер).

Скриншот без запущенного торрент клиента:
<img width="1109" alt="Снимок экрана 2024-11-04 171739" src="https://github.com/user-attachments/assets/848844a2-edc7-468b-b2b7-3eedeae068ee">
Скриншот с запущенным торрент клиентом, сразу после запуска его:
<img width="1159" alt="Снимок экрана 2024-11-04 171828" src="https://github.com/user-attachments/assets/9bd0f612-858e-4dc5-842f-ce75243896a1">
<img width="716" alt="Снимок экрана 2024-11-04 171916" src="https://github.com/user-attachments/assets/8a918e6d-ccab-4f38-88c1-2e55fcacf35c">
Может по это причине просто тупо канал забивается Adguard VPN не понятно от куда берущимся трафиком и соответственно спустя пару минут скорость соединения резко падает.
Насчет нагрузки на процессор это к задаче #693 которая так и не ушла с последней ночной сборкой.
Скриншоты приводил с версии 2.4.1 nightly 4. Локация Финляндия, проверял еще на Германия (Франкфурт) и Дания, QUIC отключен, так как с ним соединение совсем рвется, приложение отображает, что соединение есть, а если на любой сайт зайти попытаться, то выдает ошибку соединения. В торрент клиенте запущено 12 раздач активных. Такой эффект возникает когда скорость закачки торрента 50 и выше Мбит, в моем случае когда торрент клиент максимальную свою полосу разрешенную начинает использовать (ограничение в 70 Мбит максимум). И проблема всегда наблюдается только по нисходящему каналу, по восходящему скорость всегда высокая судя по результатам speedtest, даже когда в торрент клиенте восходящая скорость такая же как по нисходящему.
